compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/HeaderProcessor.kt

```diff
@@ -160,60 +166,117 @@ internal abstract class NumericForLoopHeader<T : NumericHeaderInfo>(
+                                dispatchReceiver = lastExpression
+                                putValueArgument(0, irGet(inductionVariable))
+                            })
+                            putValueArgument(1, irInt(0))
+                        }
+                    } else {
+                        irCall(elementCompFun!!).apply {
+                            putValueArgument(0, lastExpression)
+                            putValueArgument(1, irGet(inductionVariable))
+                        }
                     }
                 ProgressionDirection.INCREASING ->
-                    // inductionVar <= last (use `<` if last is exclusive)
-                    irCall(compFun).apply {
-                        putValueArgument(0, irGet(inductionVariable))
-                        putValueArgument(1, lastExpression)
+                    if (progressionType.isUnsigned) {
+                        // inductionVar <= last (use `<` if last is exclusive)
+                        irCall(intCompFun).apply {
+                            putValueArgument(0, irCall(elementCompareToFun).apply {
+                                dispatchReceiver = irGet(inductionVariable)
+                                putValueArgument(0, lastExpression)
+                            })
+                            putValueArgument(1, irInt(0))
+                        }
+                    } else {
+                        irCall(elementCompFun!!).apply {
+                            putValueArgument(0, irGet(inductionVariable))
+                            putValueArgument(1, lastExpression)
+                        }
                     }
                 ProgressionDirection.UNKNOWN -> {
                     // If the direction is unknown, we check depending on the "step" value:
                     //   // (use `<` if last is exclusive)
                     //   (step > 0 && inductionVar <= last) || (step < 0 || last <= inductionVar)
                     val stepType = progressionType.stepType(builtIns)
-                    val isLong = progressionType == ProgressionType.LONG_PROGRESSION
                     context.oror(
                         context.andand(
-                            irCall(builtIns.greaterFunByOperandType[stepType.classifierOrFail]!!).apply {
+                            irCall(builtIns.greaterFunByOperandType.getValue(stepType.classifierOrFail)).apply {
                                 putValueArgument(0, irGet(stepVariable))
-                                putValueArgument(1, if (isLong) irLong(0) else irInt(0))
+                                putValueArgument(1, if (progressionType.isLong) irLong(0) else irInt(0))
                             },
-                            irCall(compFun).apply {
-                                putValueArgument(0, irGet(inductionVariable))
-                                putValueArgument(1, lastExpression)
+                            if (progressionType.isUnsigned) {
```

Comment:
Looks like this part, and the one below, is exactly the same as in handling `DECREASING`/`INCREASING` progressions above. Maybe extract those into some functions and reuse them here?